src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java

```diff
@@ -63,42 +61,22 @@ public EmailProcessor() {
      */
     public CompletionStage<Statistics> processEmails(final TenantConfiguration tenantConfig) {
         SphereClient client = tenantConfig.getSphereClient();
-        CustomObjectQuery<JsonNode> query = CustomObjectQuery.ofJsonNode().byContainer(CONTAINER_ID).withLimit(100L)
-            .withSort(s -> s.createdAt().sort().asc());
-        return client
-            .execute(query)
-            .thenApply(response -> {
-                Statistics statistics = new Statistics(tenantConfig.getProjectKey());
-                if (response.getTotal() < 1) {
-                    LOG.info(String.format("No email to process for tenant %s", tenantConfig
-                        .getProjectKey()));
-                }
-                List<CompletableFuture<Void>> allTenants = response.getResults().stream()
-                    .map(customObject -> Optional.ofNullable(customObject)
-                        .map(CustomObject::getValue)
-                        .map(node -> node.get(EMAIL_PROPERTY_STATUS))
-                        .map(JsonNode::asText)
-                        .filter(status -> ((equalsIgnoreCase(status, STATUS_PENDING) || tenantConfig.isProcessAll())))
-                        .map(pending -> callApiEndpoint(customObject.getId(), tenantConfig)
-                            .thenAccept(statistics::update))
-                        .orElseGet(() -> {
-                            statistics.update(Statistics.RESPONSE_IGNORED);
-                            return CompletableFuture.completedFuture(null);
-                        }))
-                    .collect(toList());
+        CustomObjectQuery<JsonNode> query = CustomObjectQuery.ofJsonNode().byContainer(CONTAINER_ID);
+        if (!tenantConfig.isProcessAll()) {
+            query = query.plusPredicates(QueryPredicate.of("value(status=\"" + STATUS_PENDING + "\")"));
+        }
+        //sorting the emailobject by creationt ime, to ensure that the emails are delivered in the correct chronically
+        // order
+        query = query.withSort(s -> s.createdAt().sort().asc());
+        Statistics statistics = new Statistics(tenantConfig.getProjectKey());
 
-                // join all the email processors
-                allOf(allTenants.toArray(new CompletableFuture[0])).join();
+        final Consumer<CustomObject<JsonNode>> pageConsumer = customObject ->
+                callApiEndpoint(customObject.getId(), tenantConfig)
+                        .thenAccept(statistics::update)
+                        .toCompletableFuture().join();
 
-                client.close();
-                return statistics;
-            })
-            .exceptionally(exception -> {
```

Comment:
Is there a reason why the exception handeling is removed?  also you don't need to close the client anymore?

Reply:
01cd4477f8b2ae17befad4040e9bbd7fe2233762